slasher/db/proposer_slashings.go

```diff
@@ -0,0 +1,233 @@
+			encoded = append(encoded, v)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return toProposerSlashings(encoded)
+}
+
+// SaveProposerSlashing accepts a proposer slashing and its status header and writes it to disk.
+func (db *Store) SaveProposerSlashing(status SlashingStatus, proposerSlashing *ethpb.ProposerSlashing) error {
+	found, st, err := db.HasProposerSlashing(proposerSlashing)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if proposer slashing is already in db")
+	}
+	if found && st == status {
+		return nil
+	}
+	return db.updateProposerSlashingStatus(proposerSlashing, status)
+
+}
+
+// DeleteProposerSlashingWithStatus deletes a proposal slashing proof given its status.
+func (db *Store) DeleteProposerSlashingWithStatus(status SlashingStatus, proposerSlashing *ethpb.ProposerSlashing) error {
+	root, err := ssz.HashTreeRoot(proposerSlashing)
+	if err != nil {
+		return errors.Wrap(err, "failed to get hash root of proposerSlashing")
+	}
+	return db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(slashingBucket)
+		k := encodeStatusTypeRoot(status, SlashingType(Proposal), root)
+		if err != nil {
+			return errors.Wrap(err, "failed to get key for for proposer slashing.")
+		}
+		if err := bucket.Delete(k); err != nil {
+			return errors.Wrap(err, "failed to delete the slashing proof header from slashing bucket")
+		}
+		return nil
+	})
+}
+
+// DeleteProposerSlashing deletes a proposer slashing proof.
+func (db *Store) DeleteProposerSlashing(slashing *ethpb.ProposerSlashing) error {
+	root, err := ssz.HashTreeRoot(slashing)
+	if err != nil {
+		return errors.Wrap(err, "failed to get hash root of proposerSlashing")
+	}
+	err = db.update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(slashingBucket)
+		b.ForEach(func(k, v []byte) error {
+			if bytes.HasSuffix(k, root[:]) {
+				b.Delete(k)
+			}
+			return nil
+		})
+		return nil
+	})
+	return err
+}
+
+// HasProposerSlashing returns the slashing key if it is found in db.
+func (db *Store) HasProposerSlashing(slashing *ethpb.ProposerSlashing) (bool, SlashingStatus, error) {
+	root, err := ssz.HashTreeRoot(slashing)
+	var status SlashingStatus
+	var found bool
+	if err != nil {
+		return found, status, errors.Wrap(err, "failed to get hash root of proposerSlashing")
+	}
+	err = db.view(func(tx *bolt.Tx) error {
+		b := tx.Bucket(slashingBucket)
+		c := b.Cursor()
+		for k, _ := c.First(); k != nil; k, _ = c.Next() {
+			if bytes.HasSuffix(k, root[:]) {
+				found = true
+				status = SlashingStatus(k[0])
+				return nil
+			}
+		}
+		return nil
+	})
+	return found, status, err
+}
+
+// updateProposerSlashingStatus deletes a proposer slashing and saves it with a new status.
+// if old proposer slashing is not found a new entry is being saved as a new entry.
+func (db *Store) updateProposerSlashingStatus(slashing *ethpb.ProposerSlashing, status SlashingStatus) error {
+	root, err := ssz.HashTreeRoot(slashing)
```

Comment:
same concerns for this method, a basic CRUD operation shouldn't be doing multiple things.

Reply:
there is a complexity here of a same slashing with possible other status . i try to avoid multiple instances of the same slashing

Reply:
why does a slashing have multiple statuses ? shouldn't it only be one

Reply:
added [design doc](https://docs.google.com/document/d/16bTq9XCw1-LkdD7TSpSvyYFg_6yvV5J1mPml5U6PnXY/edit?usp=sharing) that describes why its needed. please tell me if it makes sense

Reply:
maybe you have a container with both the slashing and the status as respective fields in it ? its a bit of a wierd pattern having the status encoded in the key. And then deleting that key when you want to update the status. Or you could encode the status in the value when saving it. encodedStatus+enc . That way you avoid this deleting of keys. You can just overwrite the value saved for that particular root.